codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java

```diff
@@ -69,84 +79,9 @@
-         * currentBits will vary between 0 and 15.
-         *
-         * symbolBits is the number of bits of the the symbol being decoded, *including* all those
-         * of the parent nodes. symbolBits tells how far down the tree we are. For example, when
-         * decoding the invalid sequence {0xff, 0xff}, currentBits will be 0, but symbolBits will be
-         * 16. This is used to know if buf ended early (before consuming a whole symbol) or if
-         * there is too much padding.
-         */
-        Node node = root;
-        int current = 0;
-        int currentBits = 0;
-        int symbolBits = 0;
-        for (int i = 0; i < buf.length; i++) {
-            int b = buf[i] & 0xFF;
-            current = (current << 8) | b;
-            currentBits += 8;
-            symbolBits += 8;
-            // While there are unconsumed bits in current, keep consuming symbols.
-            while (currentBits >= 8) {
-                int c = (current >>> (currentBits - 8)) & 0xFF;
-                node = node.children[c];
-                currentBits -= node.bits;
-                if (node.isTerminal()) {
-                    if (node.symbol == HpackUtil.HUFFMAN_EOS) {
-                        throw EOS_DECODED;
-                    }
-                    baos.write(node.symbol);
-                    node = root;
-                    // Upon consuming a whole symbol, reset the symbol bits to the number of bits
-                    // left over in the byte.
-                    symbolBits = currentBits;
-                }
-            }
-        }
-
-        /*
-         * We have consumed all the bytes in buf, but haven't consumed all the symbols. We may be on
-         * a partial symbol, so consume until there is nothing left. This will loop at most 2 times.
-         */
-        while (currentBits > 0) {
-            int c = (current << (8 - currentBits)) & 0xFF;
-            node = node.children[c];
-            if (node.isTerminal() && node.bits <= currentBits) {
-                if (node.symbol == HpackUtil.HUFFMAN_EOS) {
-                    throw EOS_DECODED;
-                }
-                currentBits -= node.bits;
-                baos.write(node.symbol);
-                node = root;
-                symbolBits = currentBits;
-            } else {
-                break;
-            }
-        }
-
-        // Section 5.2. String Literal Representation
-        // A padding strictly longer than 7 bits MUST be treated as a decoding error.
-        // Padding not corresponding to the most significant bits of the code
-        // for the EOS symbol (0xFF) MUST be treated as a decoding error.
-        int mask = (1 << symbolBits) - 1;
-        if (symbolBits > 7 || (current & mask) != mask) {
-            throw INVALID_PADDING;
-        }
-
-        return baos.toByteArray();
+    public byte[] decode(ByteBuf buf, int length) throws IOException {
+        Decoder decoder = decoders.get();
```

Comment:
@normanmaurer, I had seen your commit that removes the ThreadLocal before I submitted my comment. That seems fine to prevent the discussion from getting too out-of-hand here, but I did want to note that it may be trivial to "optimize" (if anyone wanted to) because the obvious implementation may not create any garbage.